### PR TITLE
Initial observer API

### PR DESCRIPTION
**Pull request by hkshaw1990**
Previous discussion and PR : https://github.com/opentracing/opentracing-go/pull/135
Issue in opentracing-go : https://github.com/opentracing/opentracing-go/issues/136

### CODE REVIEW

File: observer.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	// Create and return a span observer. Called when a span starts.
+	// E.g :
+	//     func StartSpan(opName string, opts ...opentracing.StartSpanOption) {
+	//     var sp opentracing.Span
+	//     sso := opentracing.StartSpanOptions{}
+	//     var spObs otobserver.SpanObserver = observer.OnStartSpan(span, opName, sso)
+	//     ...
+	// }
+	// OnStartSpan function needs to be defined for a package exporting
+	// metrics as well.
```
Comment (yurishkuro) [inline observer.go]: should this comment be here?  it's a bit confusing

File: README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    SetTag()
+    Finish()
+```
+
+### Tradeoffs
+
+As noble as our thoughts might be in fetching additional metrics (other than
+latency) for a span using an observer, there are some overhead costs. Not all
+observers need to observe all the span events, in which case, we are keeping
+a (or, some) handler function(s) empty. In effect, we will still call
```
Comment (yurishkuro) [inline README.md]: maybe reword/simplify this phrase?

File: observer.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	OnStartSpan(sp Span, operationName string, options StartSpanOptions) SpanObserver
+}
+
+// SpanObserver is created by the Observer and receives notifications about
+// other Span events.
+// Client tracers should define these functions for each of the span operations
+// which should call the registered (observer) callbacks.
```
Comment (yurishkuro) [inline observer.go]: don't think this sentence is useful.

File: README.md (hunk 2 above)
Comment (hkshaw1990) [reply README.md]: @yurishkuro the new commit rephrased this and removed the other confusing comments.

**Comment by tedsuo** [comment]
Given all of the previous discussion, and the limitations we currently have due to the lack of a `SpanId` in OpenTracing, this LGTM.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+useful metrics exported out of the kernel or other metrics, profiled for a
+span. These additional metrics can help us in getting better Root-cause
+analysis. With that being said, its not just for calculation of metrics,
+it can be used for anything which needs watching the span events.
+
+## Installation and Usage
+
+The `otobserver` package provides an API. The (client) tracers (if they want
+to implement the observer interface) can implement this interface. To do
+that, first fetch the package using go get :
```
Comment (yurishkuro) [inline README.md]: this description seems a bit off. The API provided by the tracers would be something like a functional option to tracer constructor that passes an Observer. The actual observer API is implemented by a 3rd party, not by the tracer itself.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+and say :
+
+```go
+    import "github.com/opentracing-contrib/go-observer"
+```
+
+and then, define the required span event callbacks. These registered
+callbacks would then be called on span events if an observer is created.
+Multiple observers can be registered for a given distributed tracer. Have a
```
Comment (yurishkuro) [inline README.md]: nit: Tracers may allow registering multiple observers (i.e. `s/can/may/`)

File: README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    SetTag()
+    Finish()
+```
+
+### Tradeoffs
+
+As noble as our thoughts might be in fetching additional metrics (other than
+latency) for a span using an observer, there are some overhead costs. Not all
+observers need to observe all the span events, in which case, we may have
+to keep some callback function(s) empty. In effect, we will still call these
```
Comment (yurishkuro) [inline README.md]: nit: `s/function(s)/functions/`

File: observer.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2017 Hemant Kumar, IBM Corp.
+//           (c) 2017 Yuri Shkuro, Uber Technologies, Inc.
```
Comment (yurishkuro) [inline observer.go]: you don't need to include this.

**Review by yurishkuro — APPROVED** [review]

File: README.md (hunk 4 above)
Comment (hkshaw1990) [reply README.md]: Yeah. Rephrased this in the new commit.

File: README.md (hunk 5 above)
Comment (hkshaw1990) [reply README.md]: Fixed.

File: README.md (hunk 6 above)
Comment (hkshaw1990) [reply README.md]: Fixed.

File: observer.go (hunk 7 above)
Comment (hkshaw1990) [reply observer.go]: @yurishkuro Removed that from this file and added a copyright for opentracing-contrib in the LICENSE file (following the convention in other opentracing-contrib projects).
